### PR TITLE
String Splitter - Properly indent when Use Tabs is enabled

### LINKED ISSUE #40277: C# automatic string literal wrapping gets wrong indentation with Keep Tabs enabled
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/102134/c-automatic-string-literal-wrapping-gets-wrong-ind.html)._

---
I really like the feature that the C# editor now automatically adds the proper concatenation when pressing ENTER in a non-verbatim string literal, but I cannot get it to indent correctly after the second line of the string literal. I am using VS 2017 V15.3.2.

E.g. I start with a new Console Application for .NET 4.7 and type the following line inside the Main method:
```C#
			string s = "Lorem ipsum dolor sit amet";
```

Then I place the cursor after 'Lorem' and press ENTER, so I get this - still OK:
```C#
			string s = "Lorem" +
				" ipsum dolor sit amet";
```

Then I place the cursor after 'ipsum' and press ENTER, then it unexpectedly outdents by three levels:
```C#
			string s = "Lorem" +
				" ipsum" +
	" dolor sit amet";
```

For every further line break it outdents by one more level

I would be glad if this could be resolved. Or can it be caused by some editor setting?
Thank you.

---
#### Blair Wang [MSFT] on 6/26/2018, 08:43 PM: 

Thanks for reporting this problem. After careful evaluation of the problem impact &amp; severity we are closing this problem as low priority. We understand that this might not be an ideal resolution for you, but be assured that we don’t make these decisions lightly.  We prioritize problem investigations by impact and severity for a broad set of customers; the impact of this problem is limited relative to other existing problems. 
 
 We have made significant improvements to the product over the past few months, notably in the areas of performance, reliability &amp; general quality of the product. It is possible that this problem has been addressed in the latest version of Visual Studio. We encourage you to update to latest version of Visual Studio and continue to give us feedback.

#### Christian Hornung on 10/22/2019, 06:42 AM: 

The problem is still present in VS 2019 version 16.3.5, unfortunately.<br>It only occurs if the editor is set to use tabs instead of spaces for indentation - maybe that helps to narrow it down.
#### Christian Hornung on 12/4/2019, 01:45 AM: 

Still a problem in 16.4.0.

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes #40277

### CODE REVIEW

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 1]
```diff
@@ -159,9 +159,11 @@ private string GetIndentString(SyntaxNode newRoot)
 
                 var newSourceText = newDocument.GetSyntaxRootSynchronously(CancellationToken).SyntaxTree.GetText(CancellationToken);
                 var baseLine = newSourceText.Lines.GetLineFromPosition(desiredIndentation.BasePosition);
-                var baseOffsetInLine = desiredIndentation.BasePosition - baseLine.Start;
 
-                var indent = baseOffsetInLine + desiredIndentation.Offset;
+                var baseOffsetInLineInPositions = desiredIndentation.BasePosition - baseLine.Start;
+                var baseOffsetInLineInColumns = UseTabs ? baseOffsetInLineInPositions * TabSize : baseOffsetInLineInPositions;
+
+                var indent = baseOffsetInLineInColumns + desiredIndentation.Offset;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: i don't think you want this.  i think you to use GetColumnFromLineOffset

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 2]
```diff
@@ -159,9 +159,11 @@ private string GetIndentString(SyntaxNode newRoot)
 
                 var newSourceText = newDocument.GetSyntaxRootSynchronously(CancellationToken).SyntaxTree.GetText(CancellationToken);
                 var baseLine = newSourceText.Lines.GetLineFromPosition(desiredIndentation.BasePosition);
-                var baseOffsetInLine = desiredIndentation.BasePosition - baseLine.Start;
 
-                var indent = baseOffsetInLine + desiredIndentation.Offset;
+                var baseOffsetInLineInPositions = desiredIndentation.BasePosition - baseLine.Start;
+                var baseOffsetInLineInColumns = UseTabs ? baseOffsetInLineInPositions * TabSize : baseOffsetInLineInPositions;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: i don't think you want this. i think you to use GetColumnFromLineOffset

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs [hunk 3]
```diff
@@ -832,5 +836,53 @@ void M()
...
+{
+	void M()
+	{
+		var s = ""Hello "" +
+			""there "" +
+			""[||]world"";
+	}
+}",
+            useTabs: true);
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs]: for tests with tabs, i recommend having the test harness actually validate there are tabs in the before/after. to ensrue we don't accidentally lose this with an errant edit.

File: src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs [hunk 4]
```diff
@@ -159,9 +159,11 @@ private string GetIndentString(SyntaxNode newRoot)
 
                 var newSourceText = newDocument.GetSyntaxRootSynchronously(CancellationToken).SyntaxTree.GetText(CancellationToken);
                 var baseLine = newSourceText.Lines.GetLineFromPosition(desiredIndentation.BasePosition);
-                var baseOffsetInLine = desiredIndentation.BasePosition - baseLine.Start;
 
-                var indent = baseOffsetInLine + desiredIndentation.Offset;
+                var baseOffsetInLineInPositions = desiredIndentation.BasePosition - baseLine.Start;
+                var baseOffsetInLineInColumns = baseLine.GetColumnFromLineOffset(baseOffsetInLineInPositions, TabSize);
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: @CyrusNajmabadi Does this look better?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs]: yup yup!

**Comment by 06needhamt** [comment]
Should this be ported to my PR #40408 when this gets merged?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dpoeschl** [comment]
@06needhamt Feel free to rebase/merge master back into your PR, though I'm not sure why that's necessary. The code all seems unrelated to me.

**Comment by 06needhamt** [comment]
Thanks, I've implemented a comment splitter that uses the same indentation code as this PR so I will port the change into that so it doesn't suffer the same bug.

**Pull request merged**